### PR TITLE
[FLINK-25697][Runtime / Metrics] introduce port url config for prometheus push gateway a…

### PR DESCRIPTION
**Pull request by mas-chen**
…nd deprecate host/port configs

## What is the purpose of the change

Support scheme specification in Prometheus Push Gateway Reporter

## Brief change log

- Introduce new host url config that unifies scheme, host, and port specification
- Deprecate host and port configs

## Verifying this change

This change added tests and can be verified as follows:

- Added unit tests to verify the final Prometheus Push Gateway URL

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector:  no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java [hunk 1]
```diff
@@ -57,6 +60,17 @@
         this.deleteOnShutdown = deleteOnShutdown;
     }
 
+    PrometheusPushGatewayReporter(
+            String hostUrl,
+            String jobName,
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java:65]: Why do we need a second constructor and can't just change the existing one?

File: flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporterTest.java [hunk 2]
```diff
@@ -49,4 +55,52 @@ public void testParseIncompleteGroupingKey() {
...
+    @Test
+    public void testConnectToPushGatewayPreferHostUrl() {
+        PrometheusPushGatewayReporterFactory factory = new PrometheusPushGatewayReporterFactory();
+        MetricConfig metricConfig = new MetricConfig();
+        metricConfig.setProperty(HOST_URL.key(), "https://localhost:18080");
+        metricConfig.setProperty(HOST.key(), "localhost1");
+        metricConfig.setProperty(PORT.key(), "18081");
+        PrometheusPushGatewayReporter reporter = factory.createMetricReporter(metricConfig);
+        String gatewayBaseURL =
+                (String) Whitebox.getInternalState(reporter.getPushGateway(), "gatewayBaseURL");
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporterTest.java:91]: Please find a ways to test this without using mockito, e.g., by storing the URL in the reporter and making it accessible, or moving the host/port handling into a static `@VisibleForTesting` method similar to `parseGroupingKey`.

File: flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporterTest.java [hunk 3]
```diff
@@ -49,4 +55,52 @@ public void testParseIncompleteGroupingKey() {
...
+    @Test
+    public void testConnectToPushGatewayThrowsExceptionWithoutHostInformation() {
+        PrometheusPushGatewayReporterFactory factory = new PrometheusPushGatewayReporterFactory();
+        MetricConfig metricConfig = new MetricConfig();
+        Assert.assertThrows(
+                IllegalArgumentException.class, () -> factory.createMetricReporter(metricConfig));
+
+        metricConfig.setProperty(HOST.key(), "localhost");
+        Assert.assertThrows(
+                IllegalArgumentException.class, () -> factory.createMetricReporter(metricConfig));
```
Comment (zentol) [inline flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporterTest.java:104]: missing a check for it failing if only the port is set.

**Review by zentol — CHANGES_REQUESTED** [review]
The prometheus reporter documentation contains an example config that also needs an update.

File: flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java (hunk 1 above)
Comment (mas-chen) [reply flink-metrics/flink-metrics-prometheus/src/main/java/org/apache/flink/metrics/prometheus/PrometheusPushGatewayReporter.java:65]: Ah, I misinterpreted the public evolving annotation. Thought I shouldn't introduce breaking changes to this class, will clean it up.

**Comment by zentol** [comment]
@flinkbot run azure

**Review by zentol — APPROVED** [review]

**Pull request merged**